moveit_core/robot_state/src/pyrobot_state.cpp

```diff
@@ -0,0 +1,89 @@
+#include <pybind11/pybind11.h>
+#include <pybind11/numpy.h>
+#include <pybind11/stl.h>
+#include <moveit/python/pybind_rosmsg_typecasters.h>
+
+#include <moveit/robot_state/robot_state.h>
+#include <moveit/robot_state/conversions.h>
+
+namespace py = pybind11;
+using namespace robot_state;
+
+auto SET_TO_RANDOM_POSITIONS1 = py::overload_cast<>(&RobotState::setToRandomPositions);
+auto SET_TO_RANDOM_POSITIONS2 = py::overload_cast<const JointModelGroup*>(&RobotState::setToRandomPositions);
+auto SET_TO_RANDOM_POSITIONS3 = py::overload_cast<const JointModelGroup*,                 //
+                                                  random_numbers::RandomNumberGenerator&  //
+                                                  >(&RobotState::setToRandomPositions);
+
+auto SET_JOINT_GROUP_POSITIONS1 = py::overload_cast<const std::string&,         //
+                                                    const std::vector<double>&  //
+                                                    >(&RobotState::setJointGroupPositions);
+auto SET_JOINT_GROUP_POSITIONS2 = py::overload_cast<const JointModelGroup*,     //
+                                                    const std::vector<double>&  //
+                                                    >(&RobotState::setJointGroupPositions);
+
+auto SATISFIES_BOUNDS1 = py::overload_cast<const JointModelGroup*, double>(&RobotState::satisfiesBounds, py::const_);
+
+auto ROBOT_STATE_MSG_TO_ROBOT_STATE1 = py::overload_cast<const Transforms&,               //
+                                                         const moveit_msgs::RobotState&,  //
+                                                         RobotState&,                     //
+                                                         bool                             //
+                                                         >(robotStateMsgToRobotState);
+auto ROBOT_STATE_MSG_TO_ROBOT_STATE2 = py::overload_cast<const moveit_msgs::RobotState&,  //
```

Comment:
Couldn't this be handled by an appropriate type caster?

Reply:
true, although actually I think we need to somehow explicitly prevent/delete that overload. It doesn't work, and in general we should prohibit calling function that takes a ros message by mutable reference, because the mutation won't make it back to the python caller.

Reply:
Yeah, that's another concern. Generally, of course, it is possible to modify passed arguments.

Reply:
> > 1. Which python version are we supporting? Here I've written with/for python 3.8 because that's the default on 20.04 and I think most noetic users use 3.8
> 
> pybind11 is rather agnostic regarding the python version. So, why do you see a need to restrict to a specific version?
> If really needed, I would target 20.04's python version as well.
I guess I was thinking about this for any pure python code we include, which so far was just the tutorial I was working on. So actually for this PR we don't need to worry about this. I'll make a PR to the tutorials repo and there it may be relevant.

> 
> > 1. How do we generate documentation?
> 
> Do you want to have a doxygen-like API doc?

Sure, something like that. moveit_commander API docs, and those are fine. I just think we need some kind of online docs people can browse, and that we can link to in whatever tutorials may be written for using the python bindings.

Reply:
Do you know if we can  make sure this is not accidentally allowed? I thought I could just change the enable_if statement, so that the specialization didn't match if it was a ref to non-const, but I didn't work. This is what I tried:
```
struct type_caster<T, std::enable_if_t<ros::message_traits::IsMessage<T>::value &&
                                                             (std::is_const<std::remove_reference<T>>::value || !std::is_reference<T>::value)>>
```
but this still allows me to call a function that takes a ros message by ref to non-const, which is what I want to avoid.